frontend/server/src/main/java/org/pytorch/serve/ensemble/Dag.java

```diff
@@ -0,0 +1,112 @@
+            degreeMap.put("inDegree", new HashSet<String>());
+            degreeMap.put("outDegree", new HashSet<String>());
+            dagMap.put(node.getName(), degreeMap);
+        }
+    }
+
+    public boolean checkNodeExist(Node node) {
+        return nodes.containsKey(node.getName());
+    }
+
+    public boolean hasEdgeTo(Node from, Node to) {
+        return dagMap.get(from.getName()).get("inDegree").contains(to.getName());
+    }
+
+    public void addEdge(Node from, Node to) throws InvalidDAGException {
+        if (!checkNodeExist(from)) {
+            addNode(from);
+        }
+        if (!checkNodeExist(to)) {
+            addNode(to);
+        }
+
+        if (from.getName().equals(to.getName())) {
+            throw new InvalidDAGException("Self loop exception");
+        }
+
+        if (hasEdgeTo(to, from)) {
+            throw new InvalidDAGException("loop exception");
+        }
+
+        dagMap.get(from.getName()).get("outDegree").add(to.getName());
+        dagMap.get(to.getName()).get("inDegree").add(from.getName());
+    }
+
+    public Set<String> getEndNodeNames(String degree) {
+        Set<String> startNodes = new HashSet<>();
+        for (Map.Entry<String, Map<String, Set<String>>> entry : dagMap.entrySet()) {
+            Set<String> value = entry.getValue().get(degree);
+            if (value.isEmpty()) {
+                startNodes.add(entry.getKey());
+            }
+        }
+        return startNodes;
+    }
+
+    public Set<String> getStartNodeNames() {
+        return getEndNodeNames("inDegree");
+    }
+
+    public Set<String> getLeafNodeNames() {
+        return getEndNodeNames("outDegree");
+    }
+
+    public Map<String, Integer> getDegreeMap(String degree) {
+        Map<String, Integer> inDegreeMap = new HashMap<>();
+        for (Map.Entry<String, Map<String, Set<String>>> entry : dagMap.entrySet()) {
+            inDegreeMap.put(entry.getKey(), entry.getValue().get(degree).size());
+        }
+        return inDegreeMap;
+    }
+
+    public Map<String, Integer> getInDegreeMap() {
+        return getDegreeMap("inDegree");
+    }
+
+    public Map<String, Integer> getOutDegreeMap() {
+        return getDegreeMap("outDegree");
+    }
+
+    public Map<String, Node> getNodes() {
+        return nodes;
+    }
+
+    public Map<String, Map<String, Set<String>>> getDagMap() {
+        return dagMap;
+    }
+
+    public ArrayList<String> validate() throws InvalidDAGException {
+        Set<String> startNodes = getStartNodeNames();
+
+        if (startNodes.size() != 1) {
+            throw new InvalidDAGException("DAG should have only one start node");
+        }
+
+        ArrayList<String> topoSortedList = new ArrayList<>();
+        DagExecutor de = new DagExecutor(this);
+        de.execute(null, topoSortedList);
+        if (topoSortedList.size() != nodes.size()) {
+            throw new InvalidDAGException("Not a valid DAG");
+        }
+        return topoSortedList;
+    }
+}
```

Comment:
Refactor to use Guava Graph lib - https://guava.dev/releases/23.0/api/docs/com/google/common/graph/Graph.html